### PR TITLE
Enzyme main

### PR DESCRIPTION
**Pull request by cfmelend**
The initial "rough draft" for the casanovo-enzyme aware project.

### CODE REVIEW

File: casanovo/data/datasets.py [hunk 1]
```diff
@@ -231,7 +239,9 @@ def __init__(
             min_intensity=min_intensity,
             remove_precursor_tol=remove_precursor_tol,
             random_state=random_state,
+            enzyme_vocab=enzyme_vocab,
         )
+        self.stoi = {enzyme: idx for idx, enzyme in enumerate(self.enzyme_vocab)}
```
Comment (justin-a-sanders) [inline casanovo/data/datasets.py:244]: If you initialize stoi in the parent class you don't need to do it again here.

File: casanovo/denovo/dataloaders.py [hunk 2]
```diff
@@ -148,11 +157,12 @@ def _make_loader(
             collate_fn=prepare_batch,
             pin_memory=True,
             num_workers=self.n_workers,
+            shuffle=shuffle,
```
Comment (justin-a-sanders) [inline casanovo/denovo/dataloaders.py:160]: Good catch on the shuffling :)

**Review by justin-a-sanders — COMMENTED** [review]
Looks good to me functionality wise! Just left a few small style comments.

File: casanovo/data/datasets.py [hunk 3]
```diff
@@ -44,6 +44,7 @@ def __init__(
         min_intensity: float = 0.01,
         remove_precursor_tol: float = 2.0,
         random_state: Optional[int] = None,
+        enzyme_vocab: List[str] = None,
```
[the remark this answers is not in the document]
Comment (cfmelend) [reply casanovo/data/datasets.py:47]: For the prediction step we have non-annotated spectra so we need the SpectrumDataset class and the corresponding `SpectrumIndex` in depthcharge to both handle the enzyme inputs. I found this by poking around [here](https://github.com/Noble-Lab/casanovo/blob/dev/casanovo/denovo/model_runner.py#L352) in model_runner.py and [here](https://github.com/Noble-Lab/casanovo/blob/dev/casanovo/denovo/model_runner.py#L352) in dataloaders.py. Finally, [here](https://github.com/Noble-Lab/casanovo/blob/dev/casanovo/denovo/model_runner.py#L352) in the `predict()` method in model_runner.py annotated is set to false for the loaders during the predict step. I am unsure if there is a workaround here but I tested this with deleted sequence info in a very tiny batch that the model was overfit on and it correctly predicted the aa sequences.

File: casanovo/denovo/model.py [hunk 4]
```diff
@@ -139,6 +139,8 @@ def __init__(
             dropout=dropout,
             residues=residues,
             max_charge=max_charge,
+            enzyme_size=9,
```
[the remark this answers is not in the document]
Comment (cfmelend) [reply casanovo/denovo/model.py:142]: Yes it's the number of components for training. I agree and I think I should just make this a config entry. Could also read out from the config file as well and just count the number of entries in the enzymes list to avoid having to add another parameter.

File: casanovo/denovo/model.py [hunk 5]
```diff
@@ -139,6 +139,8 @@ def __init__(
...
+            padding_idx=9,
```
[the remark this answers is not in the document]
Comment (cfmelend) [reply casanovo/denovo/model.py:143]: Yes. The padding index is arbitrary but I felt it was best to have it as the "last" index of the embedding matrix and so because of the zero-indexing it could just as well be the same as the enzyme_size.

**Comment by bittremieux** [comment]
Woops, I thought this was a stale branch because the overview screen didn't show any new commits compared to `main`, but I think that's because its target branch was another derived branch. @cfmelend I don't know what the status of this is considering that the last commits are from over half a year ago, but please recreate the PR and relevant branch(es) if necessary.

**Comment by bittremieux** [comment]
Woops, I thought this was a stale branch because the overview screen didn't show any new commits compared to main, but I think that's because its target branch was another derived branch. @cfmelend I don't know what the status of this is considering that the last commits are from over half a year ago, but I've reopened the commit for now.

**Comment by bittremieux** [comment]
Woops, I thought this was a stale branch because the overview screen didn't show any new commits compared to main, but I think that's because its target branch was another derived branch. @cfmelend I don't know what the status of this is considering that the last commits are from over half a year ago, but please recreate the PR and relevant branch(es) if necessary (because I don't seem to have that capability).
